### DISCUSSION TITLE
Mirroring a repo with LFS and w/o LFS

### DISCUSSION THREAD

**Discussion by sam-cha-tfs** [git-lfs/git-lfs#5461] [Q&A]
This might be a bit of a tough question to answer. Is it possible to mirror a repo in gitlab to github and keep the github repo up to date with the gitlab repo? The gitlab repo has large files. I do not want to modify it with LFS so I created a mirror of it with lfs to github. I need help with the update part. Is this even possible since one repo has LFS while the other doesn't?

Here are my steps to create the github repo:

```
git clone <gitlab repo>
git remote add github <github url>
git lfs migrate import --everything above 100MB
git lfs push github --alll
```

Essentially I am trying to implement this but with dest repo using LFS:

https://docs.gitlab.com/ee/user/project/repository/mirror/push.html

**Comment by chrisd8088** [comment]
Well, it's certainly possible to clone a repository, convert it to using Git LFS with `git lfs migrate import`, and then push to a new remote.

I think your challenge is going to be what happens after that.  The `git lfs migrate import` is generally intended as a one-time "flag day" step, which will create a new Git history beginning with the first commit where there's a file matching the Git LFS patterns or rules you've set up.  (For instance, if you used `git lfs migrate import --everything --include="*.bin"`, it would be the first commit in history where a `*.bin` file was added; if you use `--above=100MB` instead of `--include`, it would be the first commit where a file above the given size was committed.)

In other words, the Git history is likely to diverge from your upstream remote (which doesn't use Git LFS); you'll effectively have a different repository with a different Git history, different commit SHAs, etc.

Now you can obviously push that to a new remote, no problem there.  But what do you do when a new commit is added to the original upstream?  You can't just pull down that commit because it's parent's SHA and the `HEAD` of your local repo will be different -- you no longer have a clone once you run `git lfs migrate import`; you have a new, different repo with a different set of commits.  If you try `git lfs pull` you'll get an error like `You have divergent branches and need to specify how to reconcile them.`.

Your local repo will now also have various Git LFS rules in its `.gitattributes` file, which aren't part of the upstream repo.  You could fetch from the upstream any new commits, of course, but bringing those changes into your local repo won't be able to just use the default fast-forward Git logic.  If you try to merge in the new commits, you may get `refusing to merge unrelated histories`.

If you use `git cherry-pick`, that may result in Git LFS complaining that there are files that match its `.gitattributes` which aren't Git LFS objects (e.g., `Encountered 1 file that should have been a pointer`).  You can then resolve that with `git lfs migrate import --fixup`, which will try to convert any files matching the Git LFS `.gitattributes` patterns into Git LFS objects if they aren't already.  You'd then have to repeat that process each time you fetch new commits from the upstream remote, and for each branch where there are changes.

Possibly someone else will have a better approach, and if so, I'd be interested to know about it!
